lib/config/schema.js

```diff
@@ -6,260 +6,709 @@
+  },
+  {
+    name: 'contextManager',
+    type: 'select(asynclocalstorage,asynchooks)',
+    // 'contextManager' is explicitly *not* defined in DEFAULTS because
+    // normalizeContextManager() needs to know if a value was provided by the
+    // user.
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_CONTEXT_MANAGER',
+  },
+  {
+    name: 'frameworkName',
+    type: 'string',
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_FRAMEWORK_NAME',
+  },
+  {
+    name: 'frameworkVersion',
+    type: 'string',
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_FRAMEWORK_VERSION',
+  },
+  {
+    name: 'hostname',
+    type: 'string',
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_HOSTNAME',
+  },
+  {
+    name: 'payloadLogFile',
+    type: 'string',
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_PAYLOAD_LOG_FILE',
+  },
+  {
+    name: 'serverCaCertFile',
+    type: 'string',
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_SERVER_CA_CERT_FILE',
+  },
+  {
+    name: 'secretToken',
+    type: 'string',
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_SECRET_TOKEN',
+    crossAgentName: 'secret_token',
+  },
+  {
+    name: 'serviceName',
+    type: 'string',
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_SERVICE_NAME',
+    crossAgentName: 'service_name',
+  },
+  {
+    name: 'serviceVersion',
+    type: 'string',
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_SERVICE_VERSION',
+    crossAgentName: 'service_version',
+  },
+  {
+    name: 'spanFramesMinDuration',
+    type: 'durationSecondsNegative',
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_SPAN_FRAMES_MIN_DURATION',
+    deprecated: 'Deprecated in: v3.30.0, use spanStackTraceMinDuration',
+  },
+  { name: 'ignoreUrls', type: 'stringArray', defaultValue: undefined },
+  {
+    name: 'ignoreUrlRegExp',
+    type: 'wildcardArray',
+    defaultValue: [],
+    deps: ['ignoreUrls'],
+  },
+  {
+    name: 'ignoreUrlStr',
+    type: 'wildcardArray',
+    defaultValue: [],
+    deps: ['ignoreUrls'],
+  },
+  {
+    name: 'ignoreUserAgents',
+    type: 'stringArray',
+    defaultValue: undefined,
+  },
+  {
+    name: 'ignoreUserAgentRegExp',
+    type: 'wildcardArray',
+    defaultValue: [],
+    deps: ['ignoreUserAgents'],
+  },
+  {
+    name: 'ignoreUserAgentStr',
+    type: 'wildcardArray',
+    defaultValue: [],
+    deps: ['ignoreUserAgents'],
+  },
+  // Special options that
+  // - may afect the whole config
+  // - change the behavior of thelogger
+  // - are for testing or internal use
+  {
+    name: 'configFile',
+    type: 'string',
+    defaultValue: 'elastic-apm-node.js',
+    envVar: 'ELASTIC_APM_CONFIG_FILE',
+  },
+  {
+    name: 'logger',
+    type: 'logger',
+    defaultValue: undefined,
+    envVar: 'ELASTIC_APM_LOGGER',
+  },
+  {
+    name: 'transport',
+    type: 'function',
+    defaultValue: undefined,
+    internal: true,
+  },
+  {
+    name: 'loggingPreambleData',
```

Comment:
This is a config var?

Reply:
config does the calculation of all this data in the constructor and place it into a property with that name. Then `Agent.start()` access that data to add som extra values like `activationMethod`, `dependencies` and `startTrace`.

On a second though this new schema can keep values from different sources (env, file, start) so the preamble could be generated on demand and exposed as an API. I'm going to remove it